pkg/kubelet/cm/dra/claiminfo.go

```diff
@@ -161,36 +153,76 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
 			entry.ResourceHandles,
 		)
 		for pluginName, cdiDevices := range entry.CDIDevices {
-			err := info.addCDIDevices(pluginName, cdiDevices)
+			err := info.updateCDIDevices(pluginName, cdiDevices)
 			if err != nil {
 				return nil, fmt.Errorf("failed to add CDIDevices to claimInfo %+v: %+v", info, err)
 			}
 		}
-		cache.add(info)
+		if err = cache.add(&info); err != nil {
+			return nil, fmt.Errorf("failed to add claimInfo %+v to the cache: %w", info, err)
+		}
 	}
 
 	return cache, nil
 }
 
-func (cache *claimInfoCache) add(res *ClaimInfo) {
+func (cache *claimInfoCache) add(claimInfo *ClaimInfo) error {
 	cache.Lock()
 	defer cache.Unlock()
 
-	cache.claimInfo[res.ClaimName+res.Namespace] = res
+	if _, ok := cache.claimInfo[claimInfo.ClaimName+claimInfo.Namespace]; ok {
+		return fmt.Errorf("claim %s, namespace %s: claim info already in cache", claimInfo.ClaimName, claimInfo.Namespace)
+	}
+
+	cache.claimInfo[claimInfo.ClaimName+claimInfo.Namespace] = claimInfo.DeepCopy()
+	return nil
 }
 
-func (cache *claimInfoCache) get(claimName, namespace string) *ClaimInfo {
+func (cache *claimInfoCache) update(claimName, namespace string, pluginName string, cdiDevices []string) error {
+	cache.Lock()
+	defer cache.Unlock()
+
+	claimInfo, ok := cache.claimInfo[claimName+namespace]
+	if !ok {
+		return fmt.Errorf("claim %s, namespace %s: claim info not in cache", claimName, namespace)
```

Comment:
```suggestion
		return fmt.Errorf("not present in cache")
```

Reply:
Why not to do it in one place and force all callers to repeat this? I'd prefer if callers would wrap this error and add their own details to the error.

Reply:
The convention for error messages is that the caller adds details of the variables passed in, not the callee.

Reply:
Can you point me out to the convention? It looks like golang doesn't follow it: https://go.dev/play/p/ARyRBO2gOmJ

Reply:
I can't seem to find an appropriate link at the moment. But if you think about it, there must be some convention, otherwise you end up potentially duplicating information at both levels. And the convention I always follow is -- caller includes info about passed in variables, not the callee. The caller knows more about the actual "value" of the variables than the callee, so it is in the position to add better context.

Reply:
Go is inconsistent in this regard. Some packages explicitly document that their errors include certain information, like `os` which often includes the path in errors that involved a path.

Without such documentation, the safer assumption is that errors don't include the information and thus caller should wrap errors to add it. This may duplicate information, but at least it's there.

Reply:
The guideline that I usually follow is that if the caller cannot add some piece of information, then it has to be in the error generated by the callee. Otherwise I leave it to the caller.

Adding the same piece of information to all errors returned by a function also is an indicator that this should better get added once by wrapping the different errors (less repetitive).

Reply:
I agree with @klueska that there must be some convention, but I'm not sure this project [follows](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/config.go#L64) the one that you're proposing here. Generally I somehow dislike forcing callers to repeat adding parameters to the error. In my opinion it's much safer and clearer if callee would do it once and callers would add more info from their context. May be I'm missing something obvious here. Please, enlight me :)

Reply:
The error in the callee does not necessarily stem from the use of the parameters passed in, but the caller that's wrapping that error, definitely made the call into the function with those parameters as intended inputs.

For example:
```
var users map[string]int 

func (users Users) updateUserID(username string, id int) error {
    if id < 0 {
        return fmt.Errorf("id less than zero")
    }
    if !match(username, <some pattern>) {
        return fmt.Errorf("username does not match pattern")
    }
    users[username] = id
    return nil
} 

func main() {
  username := "klueska"
  userid := 25
  if err := users.updateUserID(username userid); err != nil {
      fmt.Printf("error updating %s's userID to %d", username, userid)
      os.Exit(1)
  }
}
```

The inner function only has to return the context it knows / cares about at the time it is returning the error and the outer function makes sure to include the context it knows / cares about at the time it is returning / printing its error.

Reply:
Yes, but if your code calls updateUserID many times, you'll have to repeat adding username and id to the error messages on the caller side instead of simply using the error returned by updateUserID (e.g. "error updating klueska's userID to 25") and adding caller's info on top of it.

Reply:
Both approaches have their pros and cons. We just need to agree on what to use. It would be better to have the approach we use to be consistent with project (Kubernetes) common practices, but I don't think they exist.

Reply:
I've updated the code as you've suggested. It doesn't make sense to continue arguing further, I believe.